### ISSUE TITLE
string sql query passed to read_sql_query should be wrapped in sqlalchemy.text

### ISSUE THREAD

**Issue by haleemur** [pandas-dev/pandas#15229]
#### Code Sample, copy-pastable if proper database credentials are used

```python
import pandas as pd
from sqlalchemy import create_engine, text

# change the connection string as required
engine = create_engine('postgresql+psycopg2://hal@localhost/hal')
engine.execute('CREATE TABLE test_orders (id SERIAL PRIMARY KEY, customer_type VARCHAR)')
engine.execute("INSERT INTO test_orders (customer_type) VALUES ('urban'), ('urban'), ('rural')")

statement = "SELECT COUNT(*) FROM test_orders WHERE customer_type LIKE 'urban%'"
statement2 = "SELECT COUNT(*) FROM test_orders WHERE customer_type LIKE :c"

# does not work, exception shown in the expected output section
try: 
    df = pd.read_sql_query(statement, engine)
except Exception as e:
    print(str(type(e) + ': ' + str(e))
    

# does not work, exception shown in the expected output section
try: 
    df = pd.read_sql_query(statement, engine, params={'c': 'urban%'})
except Exception as e:
    print(str(e))

# works 
try: 
  df = pd.read_sql_query(text(statement), engine)
  df2 = pd.read_sql_query(text(statement), engine, params={'c': 'urban%'})
except Exception as e:
  print(str(e))
  
```
#### Problem description

Using plain-text sql with sqlalchemy engines don't work if 
* the `%` character is present in the query 
* with the argument `params`. 

The queries work as expected if the sql is wrapped with `sqlalchemy.text`. Also, wrapping a plain-text query with `sqlalchemy.text` should not have any effect on the output or performance of the query. 

However that is not an intuitive behavior for many analysts, who don't know what sqlalchemy is or what ORMs are in general, or that different database drivers may use `%` to format parameters.

Here's a stackoverflow Q-A on the same sqlalchemy usage issue:
http://stackoverflow.com/questions/8657508/strange-sqlalchemy-error-message-typeerror-dict-object-does-not-support-inde

We can prevent this error by wrapping the plain-text sql by `text` if an sqlalchemy engine is passed as the `con` argument to the function.

Something like this:

```python
def read_sql_query(sql, con, index_col=None, coerce_float=True, params=None, parse_dates=None, chunksize=None):
    if isinstance(sql, string) and isinstance(con, Engine):
        sql = text(sql)
    ...
```

I'd be happy to submit a PR for this code change + tests & documentation updates if the maintainers agree this to be an acceptable resolution.

#### Expected Output
```
# output from the first try-except block
<class 'TypeError'>: 'dict' object does not support indexing

# output from the second try-except block
(psycopg2.ProgrammingError) syntax error at or near ":"
LINE 1: ...ELECT COUNT(*) FROM test_orders WHERE customer_type LIKE :c1
                                                                    ^
 [SQL: 'SELECT COUNT(*) FROM test_orders WHERE customer_type LIKE :c1'] [parameters: {'c1': 'urban%'}]
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8
LOCALE: en_CA.UTF-8

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 28.8.0
Cython: 0.25.2
numpy: 1.12.0
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.1
matplotlib: 2.0.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
httplib2: None
apiclient: None
sqlalchemy: 1.1.5
pymysql: 0.7.9.None
psycopg2: 2.6.2 (dt dec pq3 ext lo64)
jinja2: 2.9.4
boto: None
pandas_datareader: None

</details>

**Comment by jorisvandenbossche** [comment]
@haleemur Thanks for the issue!

> Using plain-text sql with sqlalchemy engines don't work
>     - if the % character is present in the query
>     - with the argument params.

This does work, but the syntax you can use (https://www.python.org/dev/peps/pep-0249/#paramstyle) depends on the driver you are using. For example, psycopg2 uses the `%(name)s` style (http://initd.org/psycopg/docs/usage.html#passing-parameters-to-sql-queries). And for this reason `%` also have to be escaped as `%%`. 

But I agree this can be confusing when just using the pandas interface and the user does not really care about the driver he/she is using. 
So if wrapping the query with `text` would enable a more consistent usage pattern for the different drivers, that would certainly be welcome.

I was only wondering: if your query is wrapped in `text`, do the other paramstyle methods still work? (eg can I still use `%(c1)s`, or do I need to use `:c1`) Otherwise we would have a backwards incompatibility problem.

**Comment by haleemur** [comment]
Hmm, this does break if the user is using %-style formatting and the sqlalchemy engine.

```python
import pandas as pd
from sqlalchemy import create_engine, text
engine = create_engine('postgresql+psycopg2://hal@localhost/hal')
statement = "SELECT COUNT(*) FROM test_orders WHERE customer_type LIKE %(c1)s"
df = pd.read_sql_query(statement, engine, params={'c1': 'urban%'})
df.head()
# works fine, output is: 
#    count
# 0      2

df = pd.read_sql_query(text(statement), engine, params={'c1': 'urban%'})
# does not work, because text() escapes the ``%`` in the statement. This is the exception:
# ProgrammingError: (psycopg2.ProgrammingError) syntax error at or near "%"
# LINE 1: ...ECT COUNT(*) FROM test_orders WHERE customer_type LIKE %(c)s
#                                                                   ^
#  [SQL: 'SELECT COUNT(*) FROM test_orders WHERE customer_type LIKE %%(c)s']
```

I find it non-intuitive, because the user connects to a database through sqlalchemy engine, but has to specify parameters differently according to the underlying driver. I think a more logical behavior would be

* if user is using sqlalchemy.engine to connect: user should use `:param` style binding in the sql text.
* if user is using database connection directly: user should use binding style set by the connection in the sql text

However, this breaks compatibility, and I would prefer to maintain compatibility as much as possible.

Thoughts?

**Comment by jorisvandenbossche** [comment]
I agree that the behaviour you propose would probably have been a better default if we did that from the beginning. However, we can't just change that now as this would break almost all code that uses `params` currently without wrapping in `text()`. 
And I don't directly see a way to circumvent this ... 

Documentation about this is welcome, but of course not really a satisfying solution. It also seems we already have a little bit of documentation mentioning this: http://pandas.pydata.org/pandas-docs/stable/io.html#advanced-sqlalchemy-queries

**Comment by haleemur** [comment]
I agree with your assessment. 

1. I think the current documentation level is good, but perhaps we can link to the advanced-sqlalchemy-querys section from the documentation pages for `read_sql` & `read_sql_query`.

2. One way to enhance the user experience would be to issue a better error message when pure-text statements (optionally with params) are used with `sqlalchemy.engine`

3. Another enhancement could be to also warn the user to wrap the statement with `sqlalchemy.text` and use `:param` style binding whenever a pure-text query is used with an `sqlalchemy.engine`. 

If you agree, I can submit the PR for above.

**Comment by jorisvandenbossche** [comment]
So the idea would be, in case an error is raised by sqlalchemy, to detect that the reason is because of not using the correct bindparams syntax, and then modify the error message to put more explanation?

So eg like `pd.read_sql("SELECT * FROM table where col=:param", engine, params={'param': 1}`, that you detect that the query is not wrapped in `text()` but the `:param` style is used, and then improve the error message? 
Do I understand that correctly?

Regarding point 3, I think in general warning, even if what you do is correctly, would be a bit too intrusive.

Better linking in the docs is certainly always welcome!

**Comment by haleemur** [comment]
Hi, upon reflecting since I wrote that, I've come around to your position regarding point 3.

For point 2, yes. I'd like to handle the following error cases & improve the error message.

case 1:
 
```python
pd.read_sql("SELECT * FROM table where col LIKE 'hal%'", engine)
```

case 2:

```python
pd.read_sql("SELECT * FROM table where col=:param", engine, params={'param': 1})
``` 

As soon as I get the time I write up this PR + tests, unless someone else wants to write this.
